### PR TITLE
List of known bugs.

### PR DESCRIPTION
**Pull request by chriseth**
Adds a machine-readable list of known bugs.

Fixes https://github.com/ethereum/solidity/issues/1452

### CODE REVIEW

**Comment by axic** [comment]
I guess fixes #1452?

**Comment by axic** [comment]
Can you add an another flat list of broken version ranges?

**Comment by chriseth** [comment]
Yes! :-)

**Comment by chriseth** [comment]
The idea is to additionally create an npm package that provides a function:
`solidityBugs(usedVersion, deployedAt, optimizerUsed) -> list of potential bugs in the contract`

**Comment by chriseth** [comment]
Hm, right, or perhaps it is better to list all versions with references to the bug ids.

**Comment by chriseth** [comment]
Also added the list of "broken version ranges".

**Comment by chriseth** [comment]
This is ready for review.

**Comment by chriseth** [comment]
Test failure seems to be unrelated (emscripten bytecode upload).

File: docs/bugs.json [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        "link": "https://blog.ethereum.org/2016/11/01/security-alert-solidity-variables-can-overwritten-storage/",
+        "severity": "high",
+        "introduced": "0.1.6",
+        "fixed": "0.4.4"
+    },
+    {
+        "name": "OptimizerStaleKnowledgeAboutSHA3",
+        "summary": "The optimizer did not properly reset its knowledge about SHA3 operations resulting in some hashes (also used for storage variable positions) not being calculated correctly.",
+        "description": "The optimizer performs symbolic execution in order to save re-evaluating expressions whose value is already known. This knowledge was not properly reset across control flow paths and thus the optimizer sometimes thought that the result of a SHA3 operation is already present on the stack. This could result in data corruption by accessing the wrong storage slot.",
+        "severity": "low/medium",
```
Comment (pirapira) [inline docs/bugs.json]: Just by reading the description, I don't get the impression this can be `low`.

File: docs/bugs.rst [hunk 2]
```diff
@@ -0,0 +1,61 @@
+.. index:: Bugs
+
+.. _known_bugs:
+
+##################
+List of Known Bugs
+##################
+
+Below, you can find a JSON-formatted list of known security-relevant bugs in the
```
Comment (pirapira) [inline docs/bugs.rst]: Maybe explicitly say `some of the` known security-relevant bugs.

File: docs/bugs.rst [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+There is another file called `bugs_by_version.json
+<https://github.com/ethereum/solidity/blob/develop/docs/bugs_by_version.json>`_,
+which can be used to check which bugs affect a specific version of the compiler.
+
+Contract source verification tools and also other tools interacting with
+contracts should consult this list according to the following criteria:
+
+ - It is mildly suspicious if a contract was compiled with a nightly
+   compiler version instead of a released version. These compiler versions
+   might contain undocumented bugs.
```
Comment (pirapira) [inline docs/bugs.rst]: This sounds like, released versions do not contain undocumented bugs.  So, it's safer to remove the last sentence.

File: docs/security-considerations.rst [hunk 4]
```diff
@@ -22,7 +22,10 @@ you should be more careful.
 This section will list some pitfalls and general security recommendations but
 can, of course, never be complete. Also, keep in mind that even if your
 smart contract code is bug-free, the compiler or the platform itself might
-have a bug.
+have a bug. All known security-relevant bugs of the compiler can be found in the
```
Comment (pirapira) [inline docs/security-considerations.rst]: Maybe insert `publicly`.

Comment (pirapira) [reply docs/security-considerations.rst]: Also, I wouldn't say "all".  Any discrepancy between expectation and implementation is a bug, and we don't know people's expectations.

File: docs/bugs.json (hunk 1 above)
Comment (chriseth) [reply docs/bugs.json]: Will change it to medium.

File: docs/bugs.rst (hunk 3 above)
Comment (chriseth) [reply docs/bugs.rst]: I would like to provide some explanation at least. What about "might contain known but undocumented bugs"?

Comment (pirapira) [reply docs/bugs.rst]: Hmm.  I think `suspicious` in the first sentence is already enough, and we do not have to explain why they are suspicious.

"might contain known but undocumented bugs" would sort-of imply that all known bugs in the released versions are documented.  We should avoid giving the impression that the list is complete in any sense.

**Review by pirapira — APPROVED** [review]
Looks good to me.

**Pull request merged**